docs/proposals/20220414-runtime-hooks.md

```diff
@@ -0,0 +1,430 @@
+---
+title: Runtime Hooks for Add-on Management
+authors:
+  - "@killianmuldoon"
+  - "@ykakarap"
+reviewers:
```

Comment:
would be great to also geat a review from @jackfrancis and/or @Jont828 since they are working on the add-on story via https://docs.google.com/document/d/1TdbfXC2_Hhg0mH7-7hXcT1Gg8h6oXKrKbnJqbpFFvjw/edit#heading=h.cbd41fsf3rhi

Reply:
@jackfrancis @Jont828 Would be great if you can take a look if you have some time, thx! :)

Reply:
+1 @jackfrancis @Jont828 work could really benefit from lifecycle hook to manage edge cases like 
- addon to be upgraded before the cluster itself 
- addon to be upgraded right after the entire CP is upgraded
- addon to be upgraded only after the entire cluster is upgraded
- "blocking" cluster upgrade before an addon upgrade is completed
- addon cleanup before the cluster deletion
and probably more

Note: we are not yet there in the prototype work, but I guess we will get there soon

Reply:
Sounds good let me take a look!